lib/cache/collections.go

```diff
@@ -805,8 +805,15 @@ func (c *certAuthority) fetch(ctx context.Context) (apply func(ctx context.Conte
 		return nil, trace.Wrap(err)
 	}
 
+	// DELETE IN 11.0.
+	// missingDatabaseCA is needed only when leaf cluster v9 is connected
+	// to root cluster v10. Database CA has been added in v10, so older
+	// clusters don't have it and fetchCertAuthorities() returns an error.
+	missingDatabaseCA := false
 	applyDatabaseCAs, err := c.fetchCertAuthorities(ctx, types.DatabaseCA)
-	if err != nil {
+	if trace.IsBadParameter(err) {
+		missingDatabaseCA = true
+	} else if err != nil {
 		return nil, trace.Wrap(err)
 	}
```

Comment:
I think we should still wipe the local store - what if we connected to a v10 cluster that got downgraded? 😰
Perhaps this could be turned into an `allowBadParameter` function parameter passed to `fetchCertAuthorities` instead?

```suggestion
	if trace.IsBadParameter(err) {
		applyDatabaseCAs := func(context.Context) error {
			if err := c.trustCache.DeleteAllCertAuthorities(types.DatabaseCA); err != nil {
				if !trace.IsNotFound(err) {
					return trace.Wrap(err)
				}
			}
			return nil
		}
	}
```

Reply:
The idea behind the apply functions is that they wipe their part of the cache and then fill it up with the things that were fetched; it seems reasonable that in case the database CA suddenly goes missing (because of `BadParameter`, not because of a `NotFound`), we wipe the cache and then don't fill it up with anything.

The error from `DeleteAllCertAuthorities` is the same that's returned by the apply function in the normal case, unless I copypasted wrong 😬

Reply:
The particulars of upgrade/downgrade scenarios don't really matter here I think.  Caches must clear their previous state as part of setting up their new state, otherwise they are no longer internally consistent.  Upshot is that any fetch/apply operation which doesn't halt the cache init process _must_ clear the previous state.  If we want to treat `BadParameter` as not fatal to cache init in this case, then we must wipe the previous state.